File: zlox/src/scanner.zig
const std = @import("std");

pub const TokenType = enum(u8) {
    // Single-character tokens
    LEFT_PAREN,
    RIGHT_PAREN,
    LEFT_BRACE,
    RIGHT_BRACE,
    COMMA,
    DOT,
    MINUS,
    PLUS,
    SEMICOLON,
    SLASH,
    STAR,

    // One- or two-character tokens
    BANG,
    BANG_EQUAL,
    EQUAL,
    EQUAL_EQUAL,
    GREATER,
    GREATER_EQUAL,
    LESS,
    LESS_EQUAL,

    // Literals
    IDENTIFIER,
    STRING,
    NUMBER,

    // Keywords
    AND,
    CLASS,
    ELSE,
    FALSE,
    FOR,
    FUN,
    IF,
    NIL,
    OR,
    PRINT,
    RETURN,
    SUPER,
    THIS,
    TRUE,
    VAR,
    WHILE,

    ERROR,
    EOF,
};

pub const Token = struct {
    toktype: TokenType,
    lexeme: []const u8,
    line: u32,
};

pub const Scanner = struct {
    src: []const u8,
    start: usize,
    current: usize,
    line: u32,

    pub fn init(src: []const u8) Scanner {
        return Scanner{ .src = src, .start = 0, .current = 0, .line = 1 };
    }

    pub fn scanToken(self: *Scanner) Token {
        self.skipWhitespaceAndComments();
        self.start = self.current;
        if (self.isAtEnd()) {
            return self.makeToken(TokenType.EOF);
        }

        const c = self.advance();

        if (std.ascii.isAlphabetic(c) or c == '_') {
            return self.identifier();
        }
        if (std.ascii.isDigit(c)) {
            return self.number();
        }

        switch (c) {
            '(' => return self.makeToken(.LEFT_PAREN),
            ')' => return self.makeToken(.RIGHT_PAREN),
            '{' => return self.makeToken(.LEFT_BRACE),
            '}' => return self.makeToken(.RIGHT_BRACE),
            ';' => return self.makeToken(.SEMICOLON),
            ',' => return self.makeToken(.COMMA),
            '.' => return self.makeToken(.DOT),
            '-' => return self.makeToken(.MINUS),
            '+' => return self.makeToken(.PLUS),
            '/' => return self.makeToken(.SLASH),
            '*' => return self.makeToken(.STAR),

            '!' => return self.makeToken(if (self.match('=')) .BANG_EQUAL else .BANG),
            '=' => return self.makeToken(if (self.match('=')) .EQUAL_EQUAL else .EQUAL),
            '<' => return self.makeToken(if (self.match('=')) .LESS_EQUAL else .LESS),
            '>' => return self.makeToken(if (self.match('=')) .GREATER_EQUAL else .GREATER),

            '"' => return self.string(),

            0 => return self.makeToken(.EOF),

            else => return self.errorToken("Unexpected character."),
        }

        // return self.errorToken("Unexpected character.");
        unreachable;
    }

    fn skipWhitespaceAndComments(self: *Scanner) void {
        while (true) {
            const c = self.peek();
            switch (c) {
                ' ' => _ = self.advance(),
                '\r' => _ = self.advance(),
                '\t' => _ = self.advance(),
                '\n' => {
                    self.line += 1;
                    _ = self.advance();
                },
                '/' => {
                    if (self.peekNext() == '/') {
                        while (self.peek() != '\n' and !self.isAtEnd()) _ = self.advance();
                    } else {
                        return;
                    }
                },
                else => return,
            }
        }
    }

    fn identifier(self: *Scanner) Token {
        var c = self.peek();
        while (std.ascii.isAlphabetic(c) or std.ascii.isDigit(c) or c == '_') {
            _ = self.advance();
            c = self.peek();
        }
        return self.makeToken(self.identifierType());
    }

    fn identifierType(self: *Scanner) TokenType {
        switch (self.src[self.start]) {
            'a' => return self.checkKeyword(1, "nd", .AND),
            'c' => return self.checkKeyword(1, "lass", .CLASS),
            'e' => return self.checkKeyword(1, "lse", .ELSE),
            'f' => {
                if (self.current - self.start > 1) {
                    switch (self.src[self.start + 1]) {
                        'a' => return self.checkKeyword(2, "lse", .FALSE),
                        'o' => return self.checkKeyword(2, "r", .FOR),
                        'u' => return self.checkKeyword(2, "n", .FUN),
                        else => {},
                    }
                }
            },
            'i' => return self.checkKeyword(1, "f", .IF),
            'n' => return self.checkKeyword(1, "il", .NIL),
            'o' => return self.checkKeyword(1, "r", .OR),
            'p' => return self.checkKeyword(1, "rint", .PRINT),
            'r' => return self.checkKeyword(1, "eturn", .RETURN),
            's' => return self.checkKeyword(1, "uper", .SUPER),
            't' => {
                if (self.current - self.start > 1) {
                    switch (self.src[self.start + 1]) {
                        'h' => return self.checkKeyword(2, "is", .THIS),
                        'r' => return self.checkKeyword(2, "ue", .TRUE),
                        else => {},
                    }
                }
            },
            'v' => return self.checkKeyword(1, "ar", .VAR),
            'w' => return self.checkKeyword(1, "hile", .WHILE),

            else => {},
        }
        return .IDENTIFIER;
    }

    fn checkKeyword(self: *Scanner, offset: usize, rest: []const u8, toktype: TokenType) TokenType {
        if (self.current - self.start != offset + rest.len) {
            return .IDENTIFIER;
        }
        const sub = self.src.ptr[self.start + offset .. self.current];
        if (std.mem.eql(u8, sub, rest)) {
            return toktype;
        }
        return .IDENTIFIER;
    }

    fn number(self: *Scanner) Token {
        while (std.ascii.isDigit(self.peek())) {
            _ = self.advance();
        }
        if (self.peek() == '.' and std.ascii.isDigit(self.peekNext())) {
            _ = self.advance();
            while (std.ascii.isDigit(self.peek())) {
                _ = self.advance();
            }
        }
        return self.makeToken(TokenType.NUMBER);
    }

    fn string(self: *Scanner) Token {
        while (self.peek() != '"' and !self.isAtEnd()) {
            if (self.peek() == '\n') self.line += 1;
            _ = self.advance();
        }
        if (self.isAtEnd()) {
            return self.errorToken("Unterminated string.");
        }

        // closing quote
        _ = self.advance();
        return self.makeToken(TokenType.STRING);
    }

    fn peek(self: *Scanner) u8 {
        if (self.isAtEnd()) return 0;
        return self.src[self.current];
    }

    fn peekNext(self: *Scanner) u8 {
        if (self.isAtEnd()) return 0;
        return self.src[self.current + 1];
    }

    fn advance(self: *Scanner) u8 {
        self.current += 1;
        return self.src[self.current - 1];
    }

    fn match(self: *Scanner, expected: u8) bool {
        if (self.isAtEnd()) {
            return false;
        }
        if (self.src[self.current] != expected) {
            return false;
        }
        self.current += 1;
        return true;
    }

    fn makeToken(self: *Scanner, toktype: TokenType) Token {
        return Token{
            .toktype = toktype,
            .lexeme = if (toktype == .EOF) "\\0" else self.src[self.start..self.current],
            .line = self.line,
        };
    }

    fn errorToken(self: *Scanner, msg: []const u8) Token {
        return Token{ .toktype = TokenType.ERROR, .lexeme = msg, .line = self.line };
    }

    fn isAtEnd(self: *Scanner) bool {
        const res = self.current >= self.src.len;
        return res;
    }
};

File: rlox/src/scanner.rs
#[derive(Debug, PartialEq, Clone, Copy)]
pub enum TokenType {
    // single-character
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,

    // one- or two-character
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    // literals
    Identifier,
    String,
    Number,

    // keywords
    And,
    Class,
    Else,
    False,
    For,
    Fun,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    Error,
    EndOfFile,
}

pub struct Token {
    pub tok_type: TokenType,
    pub line: u16,
    pub lexeme: String,
}

pub struct Scanner<'a> {
    src: &'a str,
    start_idx: usize,
    current_idx: usize,
    line: u16,
}

impl<'a> Scanner<'a> {
    pub fn new(src: &'a str) -> Self {
        Scanner {
            src,
            start_idx: 0,
            current_idx: 0,
            line: 1,
        }
    }

    pub fn set_source(&mut self, src: &'a str) {
        self.src = src;
        self.start_idx = 0;
        self.current_idx = 0;
        self.line = 1;
    }

    pub fn scan_token(&mut self) -> Token {
        self.skip_whitespace_and_comments();
        self.start_idx = self.current_idx;

        if self.is_at_end() {
            return self.make_token(TokenType::EndOfFile);
        }

        let c = self.advance();

        if c.is_ascii_alphabetic() || c == '_' {
            return self.identifier();
        }

        if c.is_ascii_digit() {
            return self.number();
        }

        match c {
            '(' => self.make_token(TokenType::LeftParen),
            ')' => self.make_token(TokenType::RightParen),
            '{' => self.make_token(TokenType::LeftBrace),
            '}' => self.make_token(TokenType::RightBrace),
            ';' => self.make_token(TokenType::Semicolon),
            ',' => self.make_token(TokenType::Comma),
            '.' => self.make_token(TokenType::Dot),
            '-' => self.make_token(TokenType::Minus),
            '+' => self.make_token(TokenType::Plus),
            '/' => self.make_token(TokenType::Slash),
            '*' => self.make_token(TokenType::Star),
            '!' => {
                let tok_type = if self.mmatch('=') {
                    TokenType::BangEqual
                } else {
                    TokenType::Bang
                };
                self.make_token(tok_type)
            }
            '=' => {
                let tok_type = if self.mmatch('=') {
                    TokenType::EqualEqual
                } else {
                    TokenType::Equal
                };
                self.make_token(tok_type)
            }
            '<' => {
                let tok_type = if self.mmatch('=') {
                    TokenType::LessEqual
                } else {
                    TokenType::Less
                };
                self.make_token(tok_type)
            }
            '>' => {
                let tok_type = if self.mmatch('=') {
                    TokenType::GreaterEqual
                } else {
                    TokenType::Greater
                };
                self.make_token(tok_type)
            }
            '"' => self.string(),
            _ => self.error_token("Unexpected character."),
        }
    }

    // TODO: this does not handle UTF-8 properly, but we rollin'
    fn advance(&mut self) -> char {
        self.current_idx += 1;
        return self.src.as_bytes()[self.current_idx - 1] as char;
    }

    fn is_at_end(&self) -> bool {
        self.current_idx >= self.src.len()
    }

    fn mmatch(&mut self, expected: char) -> bool {
        if self.is_at_end() {
            return false;
        }
        if self.src.as_bytes()[self.current_idx] as char != expected {
            return false;
        }
        self.current_idx += 1;
        true
    }

    fn make_token(&self, tok_type: TokenType) -> Token {
        Token {
            tok_type,
            line: self.line,
            lexeme: self.src[self.start_idx..self.current_idx].to_string(),
        }
    }

    fn error_token(&self, msg: &str) -> Token {
        Token {
            tok_type: TokenType::Error,
            line: self.line,
            lexeme: msg.to_string(),
        }
    }

    fn skip_whitespace_and_comments(&mut self) {
        loop {
            if self.is_at_end() {
                break;
            }
            let c = self.peek();
            match c {
                ' ' | '\r' | '\t' => {
                    self.advance();
                }
                '\n' => {
                    self.line += 1;
                    self.advance();
                }
                '/' => {
                    if self.peek_next() == '/' {
                        while self.peek() != '\n' && !self.is_at_end() {
                            self.advance();
                        }
                    } else {
                        return;
                    }
                }
                _ => break,
            }
        }
    }

    fn identifier(&mut self) -> Token {
        while self.peek().is_ascii_alphabetic()
            || self.peek() == '_'
            || self.peek().is_ascii_digit()
        {
            self.advance();
        }
        self.make_token(self.identifier_type())
    }

    fn identifier_type(&self) -> TokenType {
        let c = self.src.as_bytes()[self.start_idx] as char;
        match c {
            'a' => return self.check_keyword(1, "nd", TokenType::And),
            'c' => return self.check_keyword(1, "lass", TokenType::Class),
            'e' => return self.check_keyword(1, "lse", TokenType::Else),
            'f' => {
                if self.current_idx - self.start_idx > 1 {
                    let cn = self.src.as_bytes()[self.start_idx + 1] as char;
                    match cn {
                        'a' => return self.check_keyword(2, "lse", TokenType::False),
                        'o' => return self.check_keyword(2, "r", TokenType::For),
                        'u' => return self.check_keyword(2, "n", TokenType::Fun),
                        _ => {}
                    }
                }
            }
            'i' => return self.check_keyword(1, "f", TokenType::If),
            'n' => return self.check_keyword(1, "il", TokenType::Nil),
            'o' => return self.check_keyword(1, "r", TokenType::Or),
            'p' => return self.check_keyword(1, "rint", TokenType::Print),
            'r' => return self.check_keyword(1, "eturn", TokenType::Return),
            's' => return self.check_keyword(1, "uper", TokenType::Super),
            't' => {
                if self.current_idx - self.start_idx > 1 {
                    let cn = self.src.as_bytes()[self.start_idx + 1] as char;
                    match cn {
                        'h' => return self.check_keyword(2, "is", TokenType::This),
                        'r' => return self.check_keyword(2, "ue", TokenType::True),
                        _ => {}
                    }
                }
            }
            'v' => return self.check_keyword(1, "ar", TokenType::Var),
            'w' => return self.check_keyword(1, "hile", TokenType::While),
            _ => {}
        }

        TokenType::Identifier
    }

    fn check_keyword(&self, offset: usize, rest: &str, tok_type: TokenType) -> TokenType {
        if (self.current_idx - self.start_idx) != offset + rest.len() {
            return TokenType::Identifier;
        }
        if rest == &self.src[self.start_idx + offset..self.current_idx] {
            return tok_type;
        }
        TokenType::Identifier
    }

    fn number(&mut self) -> Token {
        while !self.is_at_end() && self.peek().is_ascii_digit() {
            self.advance();
        }

        if !self.is_at_end() && self.peek() == '.' && self.peek_next().is_ascii_digit() {
            self.advance();
            while !self.is_at_end() && self.peek().is_ascii_digit() {
                self.advance();
            }
        }
        self.make_token(TokenType::Number)
    }

    fn string(&mut self) -> Token {
        while self.peek() != '"' && !self.is_at_end() {
            if self.peek() == '\n' {
                self.line += 1;
            }
            self.advance();
        }

        if self.is_at_end() {
            return self.error_token("Unterminated string.");
        }

        self.advance();
        self.make_token(TokenType::String)
    }

    fn peek(&self) -> char {
        return self.src.as_bytes()[self.current_idx] as char;
    }

    fn peek_next(&self) -> char {
        if self.is_at_end() {
            return '\0';
        }
        return self.src.as_bytes()[self.current_idx + 1] as char;
    }
}
